README.md
# Buddy Offer Component
---
React component for adding [Buddy's](https://buddy.insure) Insurance Gateway

## Requirements
Contact [Buddy](mailto:partners@buddy.insure) to obtain a `partnerID` and compatible `ions` for your specific use case.

---
## Installation
Install it from npm and include it in your React build process

`npm install @buddy-technology/offer-component`
  or 
`yarn add @buddy-technology/offer-component`

---
## Basic Usage
```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement ion="MY_ION" partnerID="my-partner-id" stage="staging" />
    </div>
  );
}
export default App;
```
---
## Props

| Option Name | Type | Default | Description
| --- | --- | --- | --- |
| *partnerID | string | | The partnerID assigned to you from Buddy |
| *ion | string | | The ion ID for the product to be presented |
| stage | enum | "STAGING" | One of: `"STAGING"`, `"PRODUCTION"`. Set to `"STAGING"` (default) for testing and development. |
| viewType | enum | `"paginated"` | Sets how the offer element is displayed. One of: `"paginated"`, `"single-form"`, `"offer-only"`. _See [View types](#view-types) for details._|
| data | object | | Customer or policy data to pre-load the offer with. Refer to specific ION documentation for the data structure.  |
| theme | object | |The [theme](#styling) object for passing in styles and color palettes |
| onUserEvent | function | | A callback for tracking user behavioral data. See [Capturing Data](#capturing-data) for details. |
| includeCheckout | boolean | `true` | Toggles whether to render the secure checkout.  |
| onAddToCart | function | | The callback fired when a user opts into an offer. See Buddy's Partner API docs for details on how to complete the transaction.|
| onRemoveCart | function | | The callback fired when a user opts out of the offer. |
_*required_

---
## View Types

Buddy's Offer Element comes in three different view types:

- __paginated__ (default) - A paginated form where users click through to the next section.
- __single form__ - A single form where all fields are displayed in the same, scrollable view.
- __offer only__ - Hide all the form fields and only display the insurance offer with the quote. Using this mode requires collecting all necessary data and a payment integration (See Buddy's [Partner API docs](https://buddyinsurance.stoplight.io/docs/partner-api-documentation/ZG9jOjE1NDc1MzQx)).

---
## Checkout
### For client-side only apps

The embed widget's default mode includes a full checkout. Customers can go through the entire flow through to a secure purchase and policies will be delivered directly to their inbox.

To explicitly set the widget to include checkout, set the `includeCheckout` property to `true`.
### For full stack apps

In server side and full-stack applications where customers are already purchasing other products, it's best to exclude the secure checkout built into the widget so customers can purchase their insurance alongside any other products as part of a single transaction. A backend integration (See Buddy's [Partner API docs](https://buddyinsurance.stoplight.io/docs/partner-api-documentation/ZG9jOjE1NDc1MzQx)) is required in order to complete the transaction.

To use this mode, set `includeCheckout` to `false`, or set viewType to `offer-only`. 

When users opt in, the `onAddToCart` callback is fired with the completed application object. If a user opts out, `onRemoveFromCart` will be called.

```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  const handleAddToCart = (payload) => {
    console.log(payload)
  };

  const handleRemoveFromCart = (payload) => {
    console.log(payload)
  };

  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement
        ion="MY_ION"
        partnerID="my-partner-id"
        viewType="offer-only"
        onAddToCart={handleAddToCart}
        onRemoveFromCart={handleRemoveFromCart}
        stage="PRODUCTION"
      />
    </div>
  );
}
export default App;
```
---
## Pre-filling Data


It's a best practice to pass in any customer and policy data that has already been collected to optimize user experience. Unless viewType is `offer-only`, users will be able to see and edit any data that is passed.

Data follows the application object structure. Refer to your specific ION (or offering) documentation.

```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  const data = {
    customer: {
      firstName: 'customer first name',
      lastName: 'customer last name',
      address: {
        line1: 'street address',
        city: 'city',
        state: 'state',
        zip: 'zip',
      }
    },
    policy: {
      startDate: '12/31/2022',
    }
  };

  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement
        ion="MY_ION"
        partnerID="my-partner-id"
        data={data}
        stage="PRODUCTION"
      />
    </div>
  );
}
export default App;
```
---
## Styling

Buddy's offer element supports extensive visual customization, allowing you to seamlessly match the design of your site with the theme prop.

1. Pick a base theme and color palette:
  Quickly get running by picking the prebuilt theme and palette that most closely resembles your website.

2. Customize even further by adding [overrides](#overrides):
   Virtually any property can be further customized using [CSS object styling](https://emotion.sh/docs/object-styles).

```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  const customTheme = {
    baseTheme: "base",
    palette: "buddy",
    overrides: {
      webFonts: [
        'https://fonts.googleapis.com/css2?family=Rubik:wght@700&display=swap',
      ],
      styles: {
        h1: {
          color: 'var(--color-text-primary)',
          fontFamily: 'Rubik, sans-serif',
          fontSize: '2rem',
          '@media (min-width: 992px)': {
            fontSize: '3rem',
          },
          '&:hover': {
            boxShadow: 'none',
            backgroundColor: '#FBF9EF',
          },
        },
        body: { color: '#0A242D' },
        '.input-text': {
          border: 'none',
          color: '#333333',
        }
      },
      colors: { textPrimary: '#0A242D' },
    },
  };

  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement
        ion="MY_ION"
        partnerID="my-partner-id"
        theme={customTheme}
        stage="PRODUCTION"
      />
    </div>
  );
}
export default App;
```
### Base Themes

We currently offer two baseThemes to start with (with more coming soon!) If you plan on doing very extensive overrides, set the baseTheme to `'none'`.

#### Available themes:

`'base`':
![base_theme.png](https://stoplight.io/api/v1/projects/cHJqOjEwMDA3NA/images/P1s10VpVT64)

`'buddy'`:
![buddy_theme.png](https://stoplight.io/api/v1/projects/cHJqOjEwMDA3NA/images/ker3veyTXrs)

### Color Palettes
We currently offer two color palettes to start with (with more coming soon!).

__Base__:

    primary: '#000000',
    secondary: '#254E70',
    tertiary: '#848484',
    negative: '#e45b78',
    positive: '#A3D9B1',
    textPrimary: '#000',
    textSecondary: '848484',
    backgroundPrimary: '#fff',
    backgroundSecondary: '#7d8387',

__Buddy__:

	primary: '#DC5648',
	secondary: '#0A242D99',
	tertiary: '#46460E33',
	negative: '#D62719',
	positive: '#A3D9B1',
	textPrimary: '#0A242D',
	textSecondary: '#0A242D99',
	backgroundPrimary: 'linear-gradient(#D6F1F2, #F8F6D2)',
	backgroundSecondary: '#FBF9EF',

### Overrides

The overrides object consists of three major properties:

`webFonts`: An array of string urls for web fonts. _Note: if using web fonts, it is still necessary to set the `fontFamily` property on any elements using that particular webfont._

`styles`: A CSS styles object with any CSS overrides on a given theme. Styles will overwrite existing theme styles. Media queries and child selectors can be accessed as a string key name, and nested within a given selector (see the example above). [Component classes](#classes-and-components) can also be manipulated within the styles object.

`colors`: A colors object will override any of the color properties in the table below:

| Key  | Where its used |
| --- | --- |
| primary | graphics, call-outs, primary buttons  |
| secondary | focused borders, accents |
| tertiary | inactive accents |
| negative | error messages (note: this is currently set to "accessible red." be mindful of accessibility when overriding this property) |
| positive | success messages |
| textPrimary | primary body copy |
| textSecondary | secondary copy, accents |
| backgroundPrimary | main background (containers) |
| backgroundSecondary | secondary backgrounds |

### Classes and Components

Some components can be directly modified by accessing their class names in the theme object. See the full list of component class names below.

```javascript
  var customTheme = {
    style: {
      h1: { fontSize: 36 },
      '.button-primary': {
        // custom styles
      },
    },
  };

```

**Available Classes**

- card
- button-primary
- button-secondary
- input-text
- input-invalid
- input-select
- input-label
- input-label-invalid
- offer-container
- field-container
- view-container

---



## Capturing data
You can hook any CRM or analytics platform into the Buddy offer-element using the `onUserEvent` hook.

`onUserEvent(eventType: string, data: object) ⇒ void`

The `onUserEvent` hook property allows you to pass in a custom function to embed your analytics of choice. Simply pass in a callback to the `onUseEvent` option, and you will get the following 2 parameters:

- **eventType:** string denoting the type of event that fired (ex: "onViewChange") 
- **data**: object with internal data regarding the event. All data objects include a timestamp property.

### Quick example

```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  const handleUserEvent = (payload) => {
    // logs an object with eventType and data to the console
    console.log(payload)
  };

  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement
        ion="MY_ION"
        partnerID="my-partner-id"
        viewType="offer-only"
        onUserEvent={handleUserEvent}
        stage="PRODUCTION"
      />
    </div>
  );
}
export default App;
```

### Event types and data fired:
We capture several user-driven events and each has different available pieces of data associated with it. Look closely at the list below to determine which events you want to capture and how to use the data.

##### `onViewChange`

Triggers when users click to a new screen in a paginated view mode.

_data returned_
| Name      | Type   | Description                                             |
|-----------|--------|---------------------------------------------------------|
| `viewId`    | string | the id of the current view                              |
| `timestamp` | number | the timestamp when the event occurred in milliseconds   |


##### `onScrollToView`

Triggers when or scroll to a section in single-page view mode.

_data returned_

| Name      | Type   | Description                                             |
|-----------|--------|---------------------------------------------------------|
| `viewId`    | string | the id of the current view                              |
| `timestamp` | number | the timestamp when the event occurred in milliseconds   |

##### `onQuote`

Triggers when the app displays retrieves a quote for the policy

_data returned_

| Name      | Type   | Description                                                                       |
|-----------|--------|-----------------------------------------------------------------------------------|
| `pricing`   | number | if successful, the price of the policy                                            |
| `error`     | string | if unsuccessful, a message explaining the type of error encountered when attempting to quote |
| `timestamp` | number | the timestamp when the event occurred in milliseconds                           |


##### `onCheckout`

Triggers during the check out process

_data returned_

| Name               | Type   | Description                                                                       |
|--------------------|--------|-----------------------------------------------------------------------------------|
| `checkoutStatus` | enum   | one of `['start', 'success', 'error']`                                               |
| `premium`        | number | the total premium of the purchase                                                 |
| `error`          | string | if unsuccessful, a message explaining the type of error encountered              |
| `timestamp`      | number | the timestamp when the event occurred in milliseconds                            |

		
##### `onFocus`

Triggers when an input is focused

_data returned_

| Name             | Type   | Description                                             |
|------------------|--------|---------------------------------------------------------|
| `elementId`    | string | the id of the invalid field/input                       |
| `timestamp`    | number | the timestamp when the event occurred in milliseconds   |


##### `onBlur`

Triggers when an input is blurred

_data returned_

| Name             | Type   | Description                                             |
|------------------|--------|---------------------------------------------------------|
| `elementId`    | string | the id of the invalid field/input                       |
| `timestamp`    | number | the timestamp when the event occurred in milliseconds   |

##### `onRadioSelection`

Triggers when a radio button is selected

_data returned_

| Name        | Type   | Description                                             |
|-------------|--------|---------------------------------------------------------|
| `elementId` | string | the id of the invalid field/input                      |
| `value`     | any    | the value of the selected radio                         |
| `timestamp` | number | the timestamp when the event occurred in milliseconds   |

##### `onSlide`

Triggers when a slider input is changed

_data returned_

| Name        | Type   | Description                                             |
|-------------|--------|---------------------------------------------------------|
| `elementId` | string | the id of the invalid field/input                      |
| `value`     | any    | the value of the selected radio                         |
| `timestamp` | number | the timestamp when the event occurred in milliseconds   |

##### `onCheckboxSelection`

Triggers when a checkbox is selected

_data returned_

| Name         | Type    | Description                                             |
|--------------|---------|---------------------------------------------------------|
| `elementId` | string  | the id of the invalid field/input                       |
| `checked`   | boolean | whether the box is checked/unchecked                    |
| `value`     | string  | the label of the selected checkbox                      |
| `timestamp` | number  | the timestamp when the event occurred in milliseconds   |

##### `onBlur`

Triggers when an input is blurred

_data returned_

| Name             | Type   | Description                                             |
|------------------|--------|---------------------------------------------------------|
| `elementId`    | string | the id of the invalid field/input                       |
| `timestamp`    | number | the timestamp when the event occurred in milliseconds   |

##### `onValidationError`

Triggers when an input is triggered as invalid

_data returned_

| Name                 | Type   | Description                                             |
|----------------------|--------|---------------------------------------------------------|
| `elementId`        | string | the id of the invalid field/input                       |
| `validationError`  | string | the type of validation error encountered                |
| `timestamp`        | number | the timestamp when the event occurred in milliseconds   |

##### `onExternalLink`

Triggers when an input is triggered as invalid

_data returned_

| Name                 | Type   | Description                                             |
|----------------------|--------|---------------------------------------------------------|
| `externalLinkUrl`  | string | the url of the link                                     |
| `timestamp`        | number | the timestamp when the event occurred in milliseconds   |

### Google Analytics example

```javascript
import React from 'react';
import BuddyOfferElement from '@buddy-technology/offer-component';

function App() {
  /* This is an example of how you could use these events
  your usage may vary depending on your analytics systems */

  const handleUserEvent = (e) => {
    // this block shows a condition looking for checkout and success, 
    // this is a completed purchase using the GA 'purchase' event for conversion
    if (e.eventType == "onCheckout" && e.data.checkoutStatus == "success") {
      gtag("event", "purchase", {
        transaction_id: e.data.orderID,
        value: e.data.premium,
        items: [{
          item_id: "SKU_12345",
          item_name: '${ion}',
          currency: "USD",
          price: e.data.premium,
        }]
      });
    }

    // This block tracks which views are displayed 
    // this is configured to use the built-in GA event 'screen_view'
    if (e.eventType == "onViewChange") {
      gtag('event', 'screen_view', {
        'app_name': 'in-offer-element',
        'screen_name': e.data.viewID
      });
    }
    // This block tracks when quotes are made
    // This is configured to use a custom event and captures a custom dimension
    if (e.eventType == "onQuote") {
      gtag('event', 'quote', {
        'app_name': 'in-offer-element',
        'price': e.data.pricing,
      });
    }
  }
  };

  return (
    <div id="app">
      <h1>My App</h1>
      <BuddyOfferElement
        ion="MY_ION"
        partnerID="my-partner-id"
        viewType="offer-only"
        onUserEvent={handleUserEvent}
        stage="PRODUCTION"
      />
    </div>
  );
}
export default App;
```

CHANGELOG.md
# CHANGELOG 

### v0.3.3  
2023-07-26  
 
- added overrideScrollToTop to options    
---  
  

### v0.3.2  
2023-06-27  
 
- fix deploy action :crossed_fingers:   
- fix includeCheckout should be optional    
---  
  

### v0.3.1  
2023-06-27  
 
- add missing props to types file   
- update docs    
---  
  

### v0.2.6  
2022-11-09  
 
- Added testing as an environment    
---  
  

### v0.2.5  
2022-08-09  
 
- remove pre-release tags    
---  
  

### v0.2.4  
2022-08-09  
 
- making public    
---  
  

### v0.2.2  
2022-07-28  
 
- added error catch for stage enums   
- few more console warnings for users   
- test-app to use DEV stage and buddytest partner    
---  
  

### v0.2.1  
2022-07-27  
 
- moved unneeded deps into devDeps and peerDeps to not cause react conflicts   
- added a catch for a non-enum stage being passed    
---  
  

### v0.1.12  
2022-05-13  
 
- Removed crossOrigin    
---  
  

### v0.1.11  
2022-05-13  
 
- The crossOrigin thing    
---  
  

### v0.1.10  
2022-05-13  
 
- Version number change    
---  
  

### v0.1.9  
2022-05-13  
 
- Added crossDomain to script tag    
---  
  

### v0.1.8  
2022-05-11  
 
- docs    
---  
  

### v0.1.7  
2022-05-11  
 
- fix build script   
- fix gh workflow   
- add changelog   
- add env urls   
- check for required props in loadScript    
---  
 

src/index.js
// eslint-disable-next-line no-unused-vars
import React, { FunctionComponent, useEffect } from 'react';
import { loadScript, updateOffer } from './util';
/**
 * @typedef {Object} DataObject
 * @property {Object} [customer] - customer data
 * @property {Object} [policy] - policy data
 */

/**
 * @typedef {Function} AddToCartFunction
 * @description A callback to be called with a payload object when users opt into the offer.
 * @returns {DataObject} the policy payload
 */

/**
 * @typedef {Function} RemoveFromCartFunction
 * @description A callback to be called with a payload object when users opt out of the offer.
 * @returns {DataObject} the policy payload
 */

/**
 * @typedef {Function} ScrollToTopFunction
 * @description A function to override the default scroll to top of offer element behavior when navigating between screens
 */

/**
 * @typedef {Object} OverridesObject
 * @property {Array<String>} [webFonts] - array of url strings linking to web fonts.
 * @property {Object} [styles] - Object Styles object for overriding any css.
 * @property {Object} [colors] - Object for overriding individual color options.
 */

/**
 * @typedef {Object} ThemeObject
 * @property {String} [baseTheme='base'] - The base theme to style with. Can be used simply as a started and overwritten.
 * @property {String} [palette='base'] - The base color scheme to style with. Can be used simply as a started and overwritten.
 * @property {OverridesObject} [overrides]
 * @example
 * const theme = {
 * 	baseTheme: 'base',
 * 		palette: 'buddy',
 * 		overrides: {
 * 			webFonts: [
 * 				'https://fonts.googleapis.com/css2?family=Rubik:wght@700&display=swap',
 * 			],
 * 			styles: {
 * 				h1: {
 * 					color: 'var(--color-text-primary)',
 * 					fontFamily: 'Rubik, sans-serif',
 * 					fontSize: '2rem',
 * 					'@media (min-width: 992px)': {
 * 						fontSize: '3rem',
 * 					},
 * 					'&:hover': {
 * 						boxShadow: 'none',
 * 						backgroundColor: '#FBF9EF',
 * 					},
 * 				},
 * 				body: { color: '#0A242D' },
 * 				'.input-text': {
 * 					border: 'none',
 * 					color: '#333333',
 * 				},
 * 			},
 * 			colors: { textPrimary: '#0A242D' },
 * 		},
 * 	};
 */

/**
 * @typedef {Object} LogoOverride
 * @property {String} url - destination url when users click the trust badge
 * @property {String} src - src of the img for displaying the trust badge
 * @property {String} [alt] - alt text for trust badge image
 */

/**
 *@typedef {Object} EventObject
 *@param {string} eventType - the type of user event (eg:'onCheckout')
 *@param {object} data - data object related to the user event.
 *@param {object} data.timestamp - timestamp of the event. All events have a timestamp property.
*/

/**
 *@typedef {Function} OnUserEventCallback
 *@param {EventObject} type - the type of user event (eg:'onCheckout')
*/

/**
 * @typedef {Object} BuddyOfferElementProps
 * @property {String} ion - The ion id for the offering.
 * @property {String} partnerID - The partner ID required for instantiating the Offer
 * @property {String} [stage="STAGING"] - toggle's the environment for the Offer Component. Defaults to STAGING. Must be set to "PRODUCTION" before going live.
 * @property {String} [viewType="paginated"] - establishes how the offer should display to the user. One of: 'paginated', 'single-form' or 'offer-only.'
 * @property {ThemeObject} [theme] - theming object for customizing offer component's styles
 * @property {DataObject} [data] - Any customer or policy data to pre-fill the offer with. Refer to your individual ION for data structure.
 * @property {OnUserEventCallback} [onUserEvent] - callback function for tracking user behavioral data. Triggers on user interactions such as input focus/blur, in-app navigation, etc. Refer to the docs for more details.
 * @property {AddToCartFunction} [onAddToCart] - callback function triggered when users opt into an offer-only offer.
 * @property {RemoveFromCartFunction} [onRemoveFromCart] - callback function triggered when users opt out of an offer-only offer.
 * @property {ScrollToTopFunction} [overrideScrollToTop] - When present, this overrides the default scroll to top of offer element behavior when navigating between screens.
 * @property {boolean} [includeCheckout] - toggles whether or not to display the card capture checkout view. Defaults to true. When false, an AddToCart callback must be provided.
 * @property {LogoOverride} [logoOverride] - object for overriding Buddy's trust badge.
 */

/**
 * @function BuddyOfferElement
 * @param {BuddyOfferElementProps} options
 * @returns {FunctionComponent<BuddyOfferElementProps>}
 */

function BuddyOfferElement(options) {
	const getScript = async () => {
		try {
			await loadScript(options);
		} catch (error) {
			// eslint-disable-next-line no-console
			console.error(error);
		}
	};
	getScript();

	useEffect(() => {
		updateOffer(options);
	}, [options]);

	return (
		<div id="buddy_offer" />
	);
}

export default BuddyOfferElement;
